index.js

```diff
@@ -42,22 +49,46 @@ const linkify = (reference, options) => {
 const domify = html => document.createRange().createContextualFragment(html);
 
 export function linkifyIssuesToHtml(string, options) {
-	options = parseOptions(options);
-	return string.replace(groupedIssueRegex, match => linkify(match, options));
+	options = applyDefaults(options);
+
+	const regex = issueRegex(options);
+	return string.replace(regex, (reference, ...matchData) => linkify(reference, matchData.at(-1), options));
 }
 
 export function linkifyIssuesToDom(string, options) {
-	options = parseOptions(options);
+	options = applyDefaults(options);
+
+	const regex = prepareRegexForSplit(issueRegex(options));
+	const parts = string.split(regex);
+	// `parts` is an array that looks like:
+	// [
+	// 	'Fixes ',                               // Initial string, or ''
+	// 	'user#123', 'user', undefined, '#123',  // Reference + capture groups
+	// 	' and ',                                // repeat
+	// 	'foo/repo#3', 'foo', 'repo', '#3',      // repeat
+	// 	'.',                                    // closing string, or ''
+	// ]
+
+	const groupsCount = countRegexGroups(regex);
 	const fragment = document.createDocumentFragment();
-	const parts = string.split(groupedIssueRegex); // All regex groups appear in the split string
-
-	for (const [index, text] of parts.entries()) {
-		if (index % issueRegexGroups === 1) { // At position issueRegexGroups n + 1 is the issue
-			fragment.append(domify(linkify(text, options)));
-		} else if (index % issueRegexGroups === 0 && text.length > 0) { // At position issueRegexGroups n + 0 is what doesn't match the regex
-			fragment.append(text);
-		}
+
+	// Iterate array one full match at a time
+	for (let index = 0; index < parts.length - 1; index += groupsCount) {
+		// Non-matching string before each reference, or ''
+		fragment.append(parts[index]);
+
+		const reference = parts[index + 1];
+		const groups = {
+			organization: parts[index + 2],
+			repository: parts[index + 3],
+			issueNumber: parts[index + 4],
+		};
+
+		fragment.append(domify(linkify(reference, groups, options)));
```

Comment:
If you don't like this, I can replace it with:

```suggestion
		fragment.append(domify(linkify(reference, regex.exec(reference).groups, options)));
```

But the loop is still more readable this way, including the `parts` explanation above.

Reply:
The current version is fine.